### ISSUE TITLE
Neomutt taking more than 7 minutes to open

### ISSUE THREAD

**Issue by vikramrajsitpal** [neomutt/neomutt#3410]
Whenever I try to open neomutt, it takes more than 5 minutes to open and after opening (if I patiently wait), does not work. [Infinite hang]

## Expected Behaviour

Just opening neomutt

## Actual Behaviour
Opening neomutt takes way too long (upwards of 7 minutes) to open up, and show the inbox. Caching of headers is not a problem either. Even when it does open up (sometimes) with the correct emails in inbox, it hangs forever. It never works and I have to terminate eventually.

https://user-images.githubusercontent.com/7771452/169736866-d2adb38a-8dec-40f7-b9ed-f95973431070.mov

## Steps to Reproduce

Just open neomutt. (screen recording above)

## How often does this happen?

- Always

## When did it start to happen?

Since I started using neotmutt after a month of not using it. It never opens up, and even when it does sometimes (after waiting patiently for too long), it immediately hangs. This never used to happen when I initially started using it in March 2022.
`top` command shows no high CPU or memory consumption.

I used brew to install it on my Mac.

I see the error sometimes which shows up in the screen recording.

## NeoMutt Version

```
NeoMutt 20220429
Copyright (C) 1996-2022 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Darwin 21.5.0 (x86_64)
ncurses: ncurses 6.3.20211021 (compiled with 6.3.20211021)
libiconv: 1.11
libidn2: 2.3.2 (compiled with 2.3.2)
GPGME: 1.17.1
OpenSSL: OpenSSL 1.1.1n  15 Mar 2022
libnotmuch: 5.6.0
storage: tokyocabinet, lmdb

Configure options: --prefix=/usr/local/Cellar/neomutt/20220429 --gss --disable-idn --idn2 --lmdb --notmuch --sasl --tokyocabinet --with-gpgme=/usr/local/opt/gpgme --with-lua=/usr/local/opt/lua --with-ssl=/usr/local/opt/openssl@1.1 --with-ui=ncurses

Compilation CFLAGS: -g -O2 -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/local/opt/gpgme/include -I/usr/local/Cellar/libassuan/2.5.5/include -I/usr/local/opt/libgpg-error/include -I/usr/local/Cellar/neomutt/20220429/include -I/usr/local/opt/lua/include/lua5.4 -DNCURSES_WIDECHAR -I/usr/local/opt/openssl@1.1/include

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  -autocrypt +fcntl -flock -fmemopen +futimens +getaddrinfo -gnutls +gpgme +gss 
  +hcache -homespool +idn -inotify -locales_hack +lua -mixmaster +nls +notmuch 
  +openssl +pgp +regex +sasl +smime -sqlite +sun_attachment 

MAILPATH="/var/mail"
PKGDATADIR="/usr/local/Cellar/neomutt/20220429/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/usr/local/Cellar/neomutt/20220429/etc"

```

## Extra Info

* Operating System and its version: macOS Monterey 12.4
* Were you using multiple copies of NeoMutt at once? No
* Were you using 'screen' or 'tmux'? No
* Is your email local (maildir) or remote (IMAP)? IMAP

**Comment by gahr** [comment]
Looks very much like an issue with your IMAP server. Can you please tell us your IMAP-relevant configurations? Also, can you please run with -d4 and post the ~/.neomuttdebug.0 file generated? Please inspect it and remove sensitive information first.

**Comment by vikramrajsitpal** [comment]
Thanks for your reply, Pietro.
### Here is my IMAP-relevant configuration:
`
set mbox_type=Maildir
set folder="imaps://imap.mail.yahoo.com:993"
set imap_user = "my_email_ID@yahoo.com"
set imap_pass = "mypassword"
set imap_check_subscribed
set imap_keepalive = 300
`

### Debug file output:
**_Please let me know in case you find something personal and I'll take it out. I removed whatever I think was personal in there._**

[neomuttdebug0.txt](https://github.com/neomutt/neomutt/files/8755850/neomuttdebug0.txt)

**Comment by gahr** [comment]
Can you please try to remove your header cache? I see `hcache uidvalidity 1340818354, uidnext 218766, modseq 0` but I also see that inbox contains `4359` messages. Perhaps something has gone out of sync badly wrt header caching. How many emails does your inbox actually contains?

**Comment by vikramrajsitpal** [comment]
Wait, something unusual happened! Neomutt started working normally again (I did not change anything)! 😮 
Maybe it was an IMAP issue from Yahoo!'s end then? I doubt that to be true tbh, but I am surprised how its all working normally all of a sudden.

**Comment by gahr** [comment]
Yeah I'd be interested of seeing the log with an empty header cache. In your latest log I still see hcache `uidvalidity 1340818354, uidnext 218774`

**Comment by gahr** [comment]
Another couple of questions: is your connection very slow? Does any other client work fine? You could also try from the command line:

1. generate your authentication token by encoding in base64 the string "username\0username\0password", substituting username and password with your actual credentials
2. talk to IMAP
```
$ openssl s_client -host imap.mail.yahoo.com -port 993 -crlf
A1 AUTHENTICATE PLAIN <the base64 string generated earlier>
A2 LSUB "" "*"
A4 SELECT "Inbox"
A5 CLOSE
ctrl-d
```

see how these commands go, what's the roundtrip time, etc.

**Comment by vikramrajsitpal** [comment]
> Another couple of questions: is your connection very slow? Does any other client work fine? You could also try from the command line:

My connection is not slow at all. Netflix's [fast.com](fast.com) shows 230 Mbps as the internet speed.

I tried talking to IMAP over command line: it was as quick as it can get.

**Comment by aabacchus** [comment]
Having built master (at this time, 41baff5151507dac2d69ff6b819a7a43a1b49805), I have also noticed a very significant increase in the startup time, since commit af90c326a3156b509cb371f6af908fd73a72a31c, which is what I was previously running. It's not a network thing (I have Maildirs). I notice in particular in .neomuttdebug:
```
[2022-06-22 22:14:51]<N> send() send: 9, 0x7fffada2e280
[2022-06-22 22:14:51]<N> email_free() NT_EMAIL_DELETE: 0x61000082a840
[2022-06-22 22:14:51]<N> notify_send() sending: NT_CONTEXT/2
[2022-06-22 22:14:51]<N> send() send: 9, 0x7fffada2e280
[2022-06-22 22:14:51]<N> email_free() NT_EMAIL_DELETE: 0x610000848440
[2022-06-22 22:14:51]<N> notify_send() sending: NT_CONTEXT/2
[2022-06-22 22:14:51]<N> send() send: 9, 0x7fffada2e280
[2022-06-22 22:14:51]<N> email_free() NT_EMAIL_DELETE: 0x6100007e3f40
[2022-06-22 22:14:51]<N> notify_send() sending: NT_CONTEXT/2
[2022-06-22 22:14:51]<N> send() send: 9, 0x7fffada2e280
```
...repeated, again and again for several *seconds*.

I deleted header cache, rebuilt it (took a similar length of time), and running again on a smaller mailbox, I notice this weird thing:
```
[2022-06-22 22:21:36]<1> mutt_parse_rc_buffer() NT_COMMAND: attachments
[2022-06-22 22:21:36]<3> parse_attach_list() added message/external-body [4]
[2022-06-22 22:21:36]<N> parse_attach_list() NT_ATTACH_ADD: message/external-body
[2022-06-22 22:21:36]<N> notify_send() sending: NT_ATTACH/1
[2022-06-22 22:21:36]<N> send() send: 4, 0
[2022-06-22 22:21:36]<N> send() send: 4, 0
[2022-06-22 22:21:36]<N> notify_send() sending: NT_COMMAND/5
[2022-06-22 22:21:36]<N> send() send: 7, 0x61f000003880
[2022-06-22 22:21:39]<2> source_rc() Reading configuration file '/home/ben/.config/neomutt/accounts/bvnf.space'
[2022-06-22 22:21:39]<1> mutt_parse_rc_buffer() NT_COMMAND: unmailboxes
```
Where did those 3 seconds go?

For an account with many more emails, the screen now shows "Sorting mailbox..." for close to 20 seconds (on commit af90c326a3156b509cb371f6af908fd73a72a31c this was ~1 second.) There are 93 commits since then.

**Comment by aabacchus** [comment]
Right, the commit which caused this increase in time is 34dc388b40666eb4d0db66596455330fc2664512. I've compiled a few and can confirm that the commit before it is good, and that one is much slower - which makes sense, if now buffers are allocated in small blocks. However, this means my problem isn't related to the problem in this issue, since the commit happened 4 days ago.

**Comment by gahr** [comment]
Can you please check if this is still the case with a fresh build off the main branch? We might have fixed this in #4357.
